src/kcas/kcas.ml

```diff
@@ -535,13 +535,17 @@ module Xt = struct
   let update0 loc f xt lt gt =
     let state = fenceless_get (as_atomic loc) in
     let before = eval state in
-    let after = f before in
-    let state =
-      if before == after && is_obstruction_free xt.casn loc then state
-      else { before; after; casn = xt.casn; awaiters = [] }
-    in
-    xt.cass <- CASN { loc; state; lt; gt; awaiters = [] };
-    before
+    match f before with
+    | after ->
+        let state =
+          if before == after && is_obstruction_free xt.casn loc then state
+          else { before; after; casn = xt.casn; awaiters = [] }
+        in
+        xt.cass <- CASN { loc; state; lt; gt; awaiters = [] };
+        before
+    | exception exn ->
+        xt.cass <- CASN { loc; state; lt; gt; awaiters = [] };
```

Comment:
This is the essential fix:  In case an exception is raised from the user given function, then the access is recorded in the transaction log as a read-only operation.

This is only necessary in the `update0` case where a location is being accessed for the first time.  If the location has already been previously accessed then it is fine to just let the exception fall through.